### ISSUE TITLE
Support multiple description markdown formats for a user minded display of a gem's 

### ISSUE THREAD

**Issue by brutuscat** [rubygems/rubygems.org#2503]
Currenly only RDoc Markup seems supported (see RubygemsHelper#simple_markup) and only detected if the first line starts with a header markup (e.g. `==`).

I figure that now that Github also has the [ability to allow developers to push packages](https://docs.github.com/en/packages/using-github-packages-with-your-projects-ecosystem/configuring-rubygems-for-use-with-github-packages), support for multiple description formats could be added and use the metadata to "communicate" what format the description is in. Then any package repository that wants to serve bundled gems should be able to render formatted HTML. Given that rubygems.org current implementation uses RDoc for converting to HTML, we could initially support what RDoc supports? That will be: _rdoc_, _markdown_, _rd_, _tomdoc_.

I was thinking using the metadata to communicate in what format the description has been written, something among this lines:

```ruby
  s.metadata = {
    'description_markup_format' => 'rdoc',
  }
```
For a list of supported formats see _RDoc::Text::MARKUP_FORMAT_

This way, given that Github and Gitlab both support Markdown and RDoc README files formats, we could have a single README file and move pieces of it to the description with minimal effort.

A second advantage will be that Rubygems.org could become a more useful or at least "formatted" front view for the gems, given that it will allow a nicer set of HTML to be displayed. I mean, it does now, but only via RDoc and other package repositories might or might not implement this the same way as ruby gems.org does... 

To sum up:
- agree on what metadata key name to use, my proposal: "description_markup_format"
- document the metadata key name and the supported formats
- make ruby gems support the standardised formats via the agreed metadata key name

Ideas? Feedback?

I'll be happy to work on this if you think it could be a good addition

**Comment by sonalkr132** [comment]
I am not particularly against this, provided that we can do this securely. 
However, if I was looking into adding support for additional formats I would probably propose using readme.md as-is from the gem file and rendering it on the version show page. Similar to what other sites are doing 
https://crates.io/crates/twitchchat
https://www.npmjs.com/package/express

As of whether you can go ahead with implementation, I would suggest waiting for more opinions on this

**Comment by brutuscat** [comment]
@sonalkr132 yes to the README part. Actually what I ended up doing in my gem was split the README in two parts to have a smaller "description" attribute. Should rubygems.org render the README instead of description?

Regarding the description itself, currently it allows rdoc or plain text, should we add a "markup descriptor" attribute as part of the metadata? I guess any gem could do this, that's why it is called "meta"... 

Maybe, following @sonalkr132  line of thought, we should then only render plain text as part of the description and rely on the README file existence convention?

**Comment by sonalkr132** [comment]
> Should rubygems.org render the README instead of description?

the description field seems more like a short intro to me. the readme can potentially have examples and basic usage, for which description is probably not suitable. I wouldn't see it is an either-or situation, we need both description and readme.

> the objectives is to make RubyGems a more useful portal for humans. Rather than having to rely always on GitHub for information about the gem. Is this something that RubyGems wants or aims to fullfil?

this a bit nuanced. rubygems.org doesn't intend to be-all and end-all place for gems related information. Our intention is to keep the versions page as simple as possible. The information in readme should ideally be served through the homepage or documentation links in the sidebar. That being said, only popular and established gems take the time to set those links and rendering readme on the versions page would make the page a bit more complete for a large number of gems.

**Comment by a-chacon** [comment]
Hi everyone! Sorry, I didn't see this issue first. But I was looking for exactly the same that @sonalkr132 says:

>the description field seems more like a short intro to me. the readme can potentially have examples and basic usage, for which description is probably not suitable. I wouldn't see it is an either-or situation, we need both description and readme.

And like others sites do the README contribute with a big support to understand what the package do.

If you think this is a good idea to implement and need help, just let me know. I like this project, and maybe I can contribute something.

**Comment by sonalkr132** [comment]
Please feel free to take this up if you are interested. Keeping the UI presentable would be tricky part imo.

**Comment by a-chacon** [comment]
Hey! @sonalkr132  this sounds like a good challenge for me. And you are right, the UI is something important, so I started trying some markdown readmes converted to html and applying some css styles for looks like rubygem's page:

[elasticsearch ](https://user-images.githubusercontent.com/39093711/172641607-7242cd48-7233-40ea-b57f-0cd16d7fe228.png)
[rainbow ](https://user-images.githubusercontent.com/39093711/172641628-9217e133-370f-48e2-aab7-3554bc1792a3.png)
[delayed_job](https://user-images.githubusercontent.com/39093711/172641648-d5678e05-ddf5-43eb-b831-cb2ce0d2c90e.png)

What do you think? Is that what you were thinking?

So after that, I need to plan how to implement this dynamically. I think there are two ways:

- Rely on the third party services. I mean GitHub has an API (Maybe GitLab too) where you can get the README in HTML ready and this 'get' could be implemented in the server side or client side, and it will be retroactive. All gems that already are in ruby gems will show his README.
- Save the README file that is uploaded in the .gem package with the version. I think this is the way, but is not retroactive, so only the versions uploaded after this implementation will show the README and I have some technical doubts here hehe

What is your opinion about this?

**Comment by simi** [comment]
Your screens look great @a-chacon. I don't think really this is the purpose of the RG.org, but if we can make it somehow simple, why not.

There are various problems we need to solve. Let me share few initial ones.

- How to get the readme?
- Currently we read only gemspec (AFAIK) during push. Do we need to extract whole gem content and scan for readme? We have already performance problems for pushing gems, we will need probably to do this in async manner.
- Should we keep README per version or is latest one enough (on your examples there is version on the top of the page, does it mean you would like to keep README per version)?

**Comment by a-chacon** [comment]
Yes @simi that is the question: How to get the readme?

I think the README is related to the version always, so yes. We should keep README per version. And to extract this from the package I don't know all the process but in the same way that is extracted the spec can be extracted the README? Like some method in Gem::Package?. And for performance you are right, this should be a job executed in async manner.

This is not a bug or a really needed feature. Like you said and sonalkr too, "I don't think really this is the purpose of the RG.org". So if this make the things worse than better for the moment (Performance), I can leave it here for now.
